src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,793 @@
+from typing import Literal
+
+import requests
+
+from github_runner_image_builder.chroot import ChrootBaseError, ChrootContextManager
+from github_runner_image_builder.config import IMAGE_OUTPUT_PATH, Arch, BaseImage
+from github_runner_image_builder.errors import (
+    BaseImageDownloadError,
+    BuildImageError,
+    DependencyInstallError,
+    ImageCompressError,
+    ImageConnectError,
+    ImageResizeError,
+    NetworkBlockDeviceError,
+    PermissionConfigurationError,
+    ResizePartitionError,
+    SystemUserConfigurationError,
+    UnattendedUpgradeDisableError,
+    UnmountBuildPathError,
+    UnsupportedArchitectureError,
+    YarnInstallError,
+    YQBuildError,
+)
+from github_runner_image_builder.utils import retry
+
+logger = logging.getLogger(__name__)
+
+SupportedBaseImageArch = Literal["amd64", "arm64"]
+
+APT_DEPENDENCIES = [
+    "qemu-utils",  # used for qemu utilities tools to build and resize image
+    "cloud-utils",  # used for growpart.
+    "golang-go",  # used to build yq from source.
+]
+APT_NONINTERACTIVE_ENV = {"DEBIAN_FRONTEND": "noninteractive"}
+SNAP_GO = "go"
+
+CHECKSUM_BUF_SIZE = 65536  # 65kb
+
+# Constants for mounting images
+IMAGE_MOUNT_DIR = Path("/mnt/ubuntu-image/")
+NETWORK_BLOCK_DEVICE_PATH = Path("/dev/nbd0")
+NETWORK_BLOCK_DEVICE_PARTITION_PATH = Path(f"{NETWORK_BLOCK_DEVICE_PATH}p1")
+
+# Constants for building image
+# This amount is the smallest increase that caters for the installations within this image.
+RESIZE_AMOUNT = "+1.5G"
+MOUNTED_RESOLV_CONF_PATH = IMAGE_MOUNT_DIR / "etc/resolv.conf"
+HOST_RESOLV_CONF_PATH = Path("/etc/resolv.conf")
+
+# Constants for disabling automatic apt updates
+APT_TIMER = "apt-daily.timer"
+APT_SVC = "apt-daily.service"
+APT_UPGRADE_TIMER = "apt-daily-upgrade.timer"
+APT_UPGRAD_SVC = "apt-daily-upgrade.service"
+
+# Constants for managing users and groups
+UBUNTU_USER = "ubuntu"
+DOCKER_GROUP = "docker"
+MICROK8S_GROUP = "microk8s"
+LXD_GROUP = "lxd"
+UBUNTU_HOME = Path("/home/ubuntu")
+
+# Constants for packages in the image
+YQ_REPOSITORY_URL = "https://github.com/mikefarah/yq.git"
+YQ_REPOSITORY_PATH = Path("yq_source")
+HOST_YQ_BIN_PATH = Path("/usr/bin/yq")
+MOUNTED_YQ_BIN_PATH = IMAGE_MOUNT_DIR / "usr/bin/yq"
+IMAGE_DEFAULT_APT_PACKAGES = [
+    "docker.io",
+    "gh",
+    "jq",
+    "npm",
+    "python3-pip",
+    "python-is-python3",
+    "shellcheck",
+    "unzip",
+    "wget",
+]
+
+
+def initialize() -> None:
+    """Configure the host machine to build images."""
+    logger.info("Installing dependencies.")
+    _install_dependencies()
+    logger.info("Enabling network block device.")
+    _enable_network_block_device()
+
+
+def _install_dependencies() -> None:
+    """Install required dependencies to run qemu image build.
+
+    Raises:
+        DependencyInstallError: If there was an error installing apt packages.
+    """
+    try:
+        output = subprocess.check_output(
+            ["/usr/bin/apt-get", "update", "-y"],
```

Comment:
There's an apt charm lib. Have you considered using it?

Reply:
Since this is was a CLI and not a charm, that library didn't quite work here - I was encountering issues with that library outside of charm context - I can't recall the exact problem though.